### ISSUE TITLE
reboot interpreter fails to write archive

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#3636]
All archiveable objects in `Archive` should "survive" a reboot of sclang. But rebooting the interpreter fails to write the archive. This is not the case when only recompiling the class library.

Reproducer:

```supercollider
Archive.at(\testing2018)

Archive.put(\testing2018, 2018);

Archive.at(\testing2018) // -> 2018

// now reboot the interpreter

Archive.at(\testing2018) // -> nil (expected: 2018)
```

**Comment by brianlheim** [comment]
I can't reproduce this (macOS 10.13).

**Comment by brianlheim** [comment]
OT - Since you are admin status in this repo, can you please make sure to label the issues and PRs you open? It helps a lot.

**Comment by patrickdupuis** [comment]
Can't reproduce on Linux (Fedora 27).

**Comment by telephon** [comment]
hm, maybe it is `macos 10.11.6`. I'll check.

>can you please make sure to label the issues and PRs you open? It helps a lot.

sure, I always label them, but usually wait to see until we know in what part of the system the problem is.

**Comment by brianlheim** [comment]
> sure, I always label them, but usually wait to see until we know in what part of the system the problem is.

OK! My bad, sorry.

**Comment by telephon** [comment]
I can still reproduce this issue using the reproducer above.

**Comment by telephon** [comment]
This is interesting, because the IDE calls exactly `0.exit` : https://github.com/supercollider/supercollider/blob/develop/editors/sc-ide/core/sc_process.cpp#L186

I don't see the difference.

**Comment by telephon** [comment]
Ok, it was a very large archive file, that failed. See my last comment on #2247.

Reproducer:

```supercollider
Archive.put(\test, \series, (0..100000000));

// now reboot interpreter
// posts: Interpreter has crashed or stopped forcefully. [Exit code: 15]

Archive.at(\test, \series); // => nil
```

**Comment by jamshark70** [comment]
Ah, interesting finding.

I think there may not be a good way to solve this. Or rather, a good way to solve it is probably more complicated than it's worth (and dev time could go elsewhere).

"Reboot interpreter" is the equivalent of MacOS "Force quit" on the sclang process. If you Force Quit something, you don't have any expectation that it will complete its shutdown process gracefully (asking to save edited documents, etc.). This is "OMG it's inflooping, I need it to stop *right now*."

AFAICS in `ScProcess::stopLanguage()` (which is called from `ScProcess::restartLanguage()`):

1. `0.exit`
2. Wait 1000 ms.
3. If sclang hasn't stopped yet, there's a call to `terminate()` (whose definition I can't find, but I'm guessing this is the "emergency stop brutally kill it" function).
4. Wait 200 ms.
5. If it's still not stopped, report the failure to the user.

If it's taking 12 seconds to write the archive, "Reboot interpreter" would naturally interrupt this process.

The problem, from the IDE's perspective, is that it can't tell the difference between the sclang interpreter being busy because of an infloop or other programmer error and being busy because of preparing a large Archive for disk writing.

If we declare that "Reboot interpreter" should wait for the interpreter to be idle, then we lose the ability to regain control in the case of an infinite loop.

So IMO we have a couple of options:

- Document that "Reboot interpreter" is an emergency stop and will not shut down cleanly.

- Or, implement an OS signal telling sclang to stop whatever is happening in the interpreter and clear all scheduler queues. Then "Reboot interpreter" could issue this signal before trying a graceful exit. The catch here is, I'm not sure anybody knows how to "stop whatever is happening in the interpreter" (and it should still handle the case where sclang fails to respond to the signal).

**Comment by jamshark70** [comment]
Related: It occurred to me recently that Archive needs a `dirty` flag. There is no need to rewrite the archive to disk on every shutdown -- if nothing changed, we should skip that step. (If you're running multiple sclang processes simultaneously, and one of them changes the archive but the others don't, those changes might be lost depending on which of the processes shuts down last. Actually we might need to merge, but... I'm still thinking about that.)

**Comment by telephon** [comment]
Indeed! Also it would already help if the `Archive` would post a message `Writing Archive …`. Then, if you get a crash message as I did, you'd guess it might have been the archive. I have now only been using reboot interpreter for more than two years (crashing the interpreter each time, but well) because recompile library was too slow :).
